presto-main/src/main/java/com/facebook/presto/server/remotetask/TaskInfoFetcher.java

```diff
@@ -280,6 +284,9 @@ public void failed(Throwable cause)
             }
             catch (RuntimeException e) {
                 onFail.accept(e);
+                // RequestErrorTracker#requestFailed throws when backoff threshold is reached. In this case we should update task info to failed state so that after
+                // recoverable execution kicks in, the query could finish properly.
+                updateTaskInfo(taskInfo.get().withTaskStatus(failWith(taskInfo.get().getTaskStatus(), FAILED, ImmutableList.of(toFailure(cause)))));
```

Comment:
Not sure if I understand the purpose of this correctly.

Reply:
Updated to 
```
// RequestErrorTracker#requestFailed throws when backoff threshold is reached.
// We should update task info to failed state, otherwise the query cannot finish because one of the task is stuck in RUNNING state.
```

Reply:
I would just say
```
// onFail callback does not update task info to failed state. 
// With recoverable grouped execution, a single failed task doesn't fail the whole query. Not updating the task info make query not finish in tests because failed task is stuck in RUNNING state.
// TODO: This issue only happens in test TestHiveRecoverableGroupedExecution when a worker is closed. Investigate why this is not an issue for production query recovery.
```

Reply:
What about the following:
```
// onFail callback does not update task info to failed state. 
// With recoverable grouped execution, a single failed task doesn't fail the whole query. Not updating the task info make query not finish in tests because failed task is stuck in RUNNING state.
// TODO: This issue only happens in test TestHiveRecoverableGroupedExecution when a worker is closed. Investigate why this is not an issue for production query recovery.
```

This explains what is missing from `onFail`, and why it's now a problem (since we don't fail query now). 

"RequestErrorTracker#requestFailed throws when backoff threshold is reached." explains why the exception will be thrown, in my opinion is a low level detail. The key here is the task somehow failed, but the task info is not updated. It's not a problem before since the query will fail. But now it can make the task stuck :/